iothub/device/src/InternalClient.cs

```diff
@@ -1468,11 +1402,7 @@ private async Task OnDeviceMessageReceivedAsync(Message message)
 
             try
             {
-                // Wait to acquire the _deviceReceiveMessageSemaphore. This ensures that we get a reference to the latest callback set on the device client.
-                await _deviceReceiveMessageSemaphore.WaitAsync().ConfigureAwait(false);
                 callbackContextTuple = _deviceReceiveMessageCallback;
-                _deviceReceiveMessageSemaphore.Release();
```

Comment:
This change is a bit orthogonal, but there was no reason to use a semaphore here. It reads from a shared resource, but it doesn't write to the shared resource so there is no concurrency risk

Reply:
What about the scenario where a user supplies a different callback right before the client was about to actually invoke it?

Reply:
The callback can be updated immediately after this value read even with this lock. Even with the semaphore, there is no guarantee that the read callback will still be valid after a millisecond. Like I mentioned on the other comment, I can remove this change since it isn't critical to fix the github issue in question, but we should still consider removing it at some point

Reply:
Looking at this more If we're trying to guarantee that the callback is guaranteed we'd need to do the following. If we don't care then Tim is correct the Semaphore is not helping here.

1. Initialize a new tuple so we get the correct references
2. Move the lock to after the callback finished.

```
            try
            {
                // Wait to acquire the _deviceReceiveMessageSemaphore. This ensures that we get a reference to the latest callback set on the device client.
                await _deviceReceiveMessageSemaphore.WaitAsync().ConfigureAwait(false);
                Tuple<ReceiveMessageCallback, object> callbackContextTuple = new Tuple<ReceiveMessageCallback, object>(_deviceReceiveMessageCallback.Item1, _deviceReceiveMessageCallback.Item2);
                _deviceReceiveMessageSemaphore.Release();

                await (callbackContextTuple?.Item1?.Invoke(message, callbackContextTuple.Item2)
                    ?? TaskHelpers.CompletedTask).ConfigureAwait(false);
            }
            finally
            {
                if (Logging.IsEnabled)
                {
                    Logging.Exit(this, message, nameof(OnDeviceMessageReceivedAsync));
                }
            }
```

Reply:
Yep. Since I'm getting some pushback on removing the semaphore grab here, I did what you suggested above, but without declaring the tuple. Now it is 

```csharp
            await _deviceReceiveMessageSemaphore.WaitAsync().ConfigureAwait(false);
            try
            {
                await (_deviceReceiveMessageCallback?.Item1?.Invoke(message, _deviceReceiveMessageCallback.Item2)
                    ?? TaskHelpers.CompletedTask).ConfigureAwait(false);
            }
            finally
            {
                _deviceReceiveMessageSemaphore.Release();
             
                if (Logging.IsEnabled)
                {
                    Logging.Exit(this, message, nameof(OnDeviceMessageReceivedAsync));
                }
            }
```

Reply:
Now the lock is a bit more useful since it makes sure that the two calls to get values from the callback (.Item1 and .Item2) will come from the same callback instance. Without the lock, there would be a chance that the callback value is set inbetween the retrieval of item1 and item2

Reply:
Not sure I like this since we will block the semaphore until method is complete. Let's get a reference to the two items from the Tuple during the Semaphore operation in the Semaphore context and then invoke the callback after we've released the lock.

Reply:
(code written not compiled)

```C#
await _deviceReceiveMessageSemaphore.WaitAsync().ConfigureAwait(false);
var item1 = _deviceReceiveMessageCallback?.Item1;
var item2 = _deviceReceiveMessageCallback?.Item2;
_deviceReceiveMessageSemaphore.Release();
            try
            {
                await (item1?.Invoke(message, item2)
                    ?? TaskHelpers.CompletedTask).ConfigureAwait(false);
            }
            finally
            {            
                if (Logging.IsEnabled)
                {
                    Logging.Exit(this, message, nameof(OnDeviceMessageReceivedAsync));
                }
            }
```

Reply:
Good point. Callbacks can take a pretty variable amount of time

Reply:
Wait, how is this different from the code that we had before?

Reply:
There isn't a need for us to await the completion of this user supplied callback, we can simply invoke it and then release the semaphore.

Reply:
As of now, the code behaves the same way as it used to. I just refactored it a bit for readability